### PR TITLE
Validate skiplist nodes in seek operations

### PR DESCRIPTION
**Pull request by JiayuZzz**
### What problem does this PR solve?

Problem Summary:

pmem record of skiplist maybe updated during seek, so we need to validate them

### What is changed and how it works?

What's Changed:

- catch exceptions for key comparation
- check validation of skiplist dram/pmem node after comparing keys

### Check List 

Tests 

- Unit test
- stress test

Side effects

- Performance regression
   No significant performance difference, <1% insert performance decrease on skiplist insert

### CODE REVIEW

**Comment by JiayuZzz** [comment]
ABA problem will be resolved in another patch

File: engine/skiplist.cpp [hunk 1]
```diff
@@ -60,17 +60,28 @@ void SkiplistNode::SeekNode(const pmem::obj::string_view &key,
...
         // will be roll back in next round
         continue;
       }
-      int cmp = compare_string_view(key, next->UserKey());
+
+      DLRecord *next_pmem_record = next->record;
+      // pmem record maybe updated during seek, then the next record could be
+      // invalid, so we need to catch exceptions here
+      int cmp;
+      try {
```
Comment (sherry-1001) [inline engine/skiplist.cpp:75]: why need `catch`?

File: engine/skiplist.cpp [hunk 2]
```diff
@@ -60,17 +60,28 @@ void SkiplistNode::SeekNode(const pmem::obj::string_view &key,
...
+        cmp = compare_string_view(key, next->UserKey());
+      } catch (std::runtime_error &) {
+        continue;
+      }
+      // pmem record maybe updated before comparing string, then the compare
```
Comment (sherry-1001) [inline engine/skiplist.cpp:80]: before? or after? if before, maybe move  line 75 ~ line 79 to after‘ `double check`

File: engine/skiplist.cpp (hunk 1 above)
Comment (JiayuZzz) [reply engine/skiplist.cpp:75]: because the next_pmem_record maybe modified, then its key maybe a invalid address

Comment (JiayuZzz) [reply engine/skiplist.cpp:75]: because the next_pmem_record maybe modified, then its key maybe a invalid address

File: engine/skiplist.cpp (hunk 2 above)
Comment (JiayuZzz) [reply engine/skiplist.cpp:80]: before or during, so we need to double check after compare, to ensure it's never changed

File: engine/skiplist.cpp (hunk 1 above)
Comment (sherry-1001) [reply engine/skiplist.cpp:75]: Considering a situation： we can access the invalid address, but key is strange string,  this code can't catch this situation.

Comment (JiayuZzz) [reply engine/skiplist.cpp:75]: This can not pass double check, the dram node will not point to it

Comment (JiayuZzz) [reply engine/skiplist.cpp:75]: This can not pass double check, the dram node will not point to it

Comment (sherry-1001) [reply engine/skiplist.cpp:75]: I think we only need double check,  `catch` seems no need

File: engine/skiplist.hpp [hunk 3]
```diff
@@ -341,6 +341,18 @@ class Skiplist : public PersistentList {
...
   }
 
+  bool ValidateDLRecord(const DLRecord *record) {
+    // if record is not a valid dl record, then its prev maybe a invalid
+    // pointer, so we need to catch exception here
+    try {
+      DLRecord *prev = pmem_allocator_->offset2addr<DLRecord>(record->prev);
+      return prev->next == pmem_allocator_->addr2offset(record) &&
+             SkiplistId(record) == id_;
+    } catch (std::runtime_error &e) {
```
Comment (ZiyanShi) [inline engine/skiplist.hpp:351]: What runtime_error will be catched? Does any function in this function throws an standard library exception?

File: engine/skiplist.cpp [hunk 4]
```diff
@@ -57,17 +58,28 @@ void SkiplistNode::SeekNode(const StringView &key, uint8_t start_height,
...
-      int cmp = compare_string_view(key, next->UserKey());
+
+      DLRecord *next_pmem_record = next->record;
+      // pmem record maybe updated during seek, then the next record could be
+      // invalid, so we need to catch exceptions here
+      int cmp;
+      try {
+        cmp = compare_string_view(key, next->UserKey());
+      } catch (std::runtime_error &) {
+        continue;
```
Comment (ZiyanShi) [inline engine/skiplist.cpp:76]: What's caught here?

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

**Pull request closed without merging**